definitions/npm/firebase_v4.x.x/flow_v0.46.x-/firebase_v4.x.x.js

```diff
@@ -0,0 +1,410 @@
+  delete(): Promise<void>;
+  getIdToken(forceRefresh?: boolean): Promise<string>;
+  getToken(forceRefresh?: boolean): Promise<string>;
+  linkAndRetrieveDataWithCredential(credential: AuthCredential): Promise<UserCredential>;
+  linkWithCredential(credential: AuthCredential): Promise<User>;
+  linkWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  linkWithPopup(provider: OAuthProvider): Promise<UserCredential>;
+  reauthenticateAndRetrieveDataWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  reload(): Promise<void>;
+  sendEmailVerification(): Promise<void>;
+  toJSON(): Object;
+  unlink(providerId: string): Promise<User>;
+  updateEmail(newEmail: string): Promise<void>;
+  updatePassword(newPassword: string): Promise<void>;
+  updatePhoneNumber(phoneCredential: AuthCredential): Promise<void>;
+  updateProfile(profile: { displayName: ?string, photoURL: ?string }): Promise<void>;
+}
+
+declare class EmailAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  providerId: string;
+  credential(email: string, password: string): AuthCredential;
+}
+
+declare class FacebookAuthProvider extends AuthProvider  {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): FacebookAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): FacebookAuthProvider;
+}
+
+declare class GithubAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): GithubAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GithubAuthProvider;
+}
+
+declare class GoogleAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(idToken?: string, accessToken?: string): AuthCredential;
+  addScope(scope: string): GoogleAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GoogleAuthProvider;
+}
+
+declare class PhoneAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  constructor(auth?: Auth): PhoneAuthProvider;
+  credential(verificationId: string, verificationCode: string): AuthCredential;
+  verifyPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<string>;
+}
+
+declare class TwitterAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string, secret: string): AuthCredential;
+  setCustomParameters(customOAuthParameters: Object): this;
+}
+
+declare type OAuthProvider = | FacebookAuthProvider
+                             | GithubAuthProvider
+                             | GoogleAuthProvider
+                             | TwitterAuthProvider;
+
+/****** database ******/
+declare type FirebaseValue = {} | Array<FirebaseValue> | string | number | boolean | null;
```

Comment:
I'm on my phone right now but hopefully this makes sense. I think the main problem I have with this is if I had a variable that had the type Workbook = { connection: { label: string }. Now if I wanted to call firebase.database.ref.update flow will now say it's an error that my type Workbook does not match FirebasValue (thought it's clearly an object). Another issue is if I wanted to assign this.workbook = snapshot.val() flow wouldn't let me even if I decided to try to check if snapshot.val instance Object and then check if it had a property connection and then check if label was a string. To get it to stop giving me errors I had to cast it to type any and then cast it to the type I declared. I think it might better to let the developer use their own types. Though if you have good examples on how you use the FirebaseValue I wouldn't mind updating my code. I'm still kind of new to flow so let me know if I'm doing anything wrong.